<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BiocSIMBA - graph-based embedding of multiomic outcomes • BiocSIMBA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="BiocSIMBA - graph-based embedding of multiomic outcomes">
<meta property="og:description" content="BiocSIMBA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BiocSIMBA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/BiocSIMBA.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vjcitn/BiocSIMBA/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="BiocSIMBA_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>BiocSIMBA - graph-based embedding of multiomic outcomes</h1>
                        <h4 data-toc-skip class="author">Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">November 13, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/BiocSIMBA/blob/HEAD/vignettes/BiocSIMBA.Rmd"><code>vignettes/BiocSIMBA.Rmd</code></a></small>
      <div class="hidden name"><code>BiocSIMBA.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>SIMBA (<a href="https://www.nature.com/articles/s41592-023-01899-8" class="external-link">paper link</a>) (<a href="https://github.com/pinellolab/simba" class="external-link">github</a>) provides methods for “graph embeddings” of relationships between elements assayed in multimodal single cell experiments. The main software stack is written in Python. This package uses basilisk to simplify use of the SIMBA methods in R/Bioconductor.</p>
<p>Steps that can be taken with this software:</p>
<ul>
<li>preprocess an RNA-seq experiment serialized as AnnData/h5ad, using scanpy methods</li>
<li>discretize the expression measures</li>
<li>form a typed graph with weighted edges suitable for embedding computations by a customized version of pytorch-biggraph</li>
<li>visualize the graph-based joint embedding of genes and cells</li>
</ul>
</div>
<div class="section level2">
<h2 id="illustration">Illustration<a class="anchor" aria-label="anchor" href="#illustration"></a>
</h2>
<p>The TENx 3k PBMC data are included in the package.</p>
<p>The heart of the system is <code>gen_graph_sce</code>. This encapsulates calls from R to python and currently employs default settings for many tuning parameters. The <code>simba_config</code> function can be used to modify some parameters, but this needs more work as of 13.11.23.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/BiocSIMBA">BiocSIMBA</a></span><span class="op">)</span></span>
<span><span class="va">h5adpath</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"h5ad"</span>, <span class="st">"tenx3k.h5ad"</span><span class="op">)</span>, package<span class="op">=</span><span class="st">"BiocSIMBA"</span><span class="op">)</span></span>
<span><span class="va">gout</span> <span class="op">=</span> <span class="fu"><a href="../reference/gen_graph_sce.html">gen_graph_sce</a></span><span class="op">(</span><span class="va">h5adpath</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Saving results in: /home/vincent/TEMP/RtmpXoctuO/file326126f5c4253/simba_work</span></span>
<span><span class="co">## relation0: source: C, destination: G</span></span>
<span><span class="co">## #edges: 450343</span></span>
<span><span class="co">## relation1: source: C, destination: G</span></span>
<span><span class="co">## #edges: 1033680</span></span>
<span><span class="co">## relation2: source: C, destination: G</span></span>
<span><span class="co">## #edges: 440524</span></span>
<span><span class="co">## relation3: source: C, destination: G</span></span>
<span><span class="co">## #edges: 227588</span></span>
<span><span class="co">## relation4: source: C, destination: G</span></span>
<span><span class="co">## #edges: 134749</span></span>
<span><span class="co">## Total number of edges: 2286884</span></span>
<span><span class="co">## Writing graph file "pbg_graph.txt" to "/home/vincent/TEMP/RtmpXoctuO/file326126f5c4253/simba_work/pbg/graph0" ...</span></span>
<span><span class="co">## Finished.</span></span>
<span><span class="co">## Auto-estimating weight decay ...</span></span>
<span><span class="co">## `.settings.pbg_params['wd']` has been updated to 0.015495</span></span>
<span><span class="co">## Weight decay being used for training is 0.015495</span></span>
<span><span class="co">## Converting input data ...</span></span>
<span><span class="co">## [2023-11-13 08:07:58.769330] Using the 5 relation types given in the config</span></span>
<span><span class="co">## [2023-11-13 08:07:58.769434] Searching for the entities in the edge files...</span></span>
<span><span class="co">## [2023-11-13 08:08:00.601742] Entity type C:</span></span>
<span><span class="co">## [2023-11-13 08:08:00.601927] - Found 2700 entities</span></span>
<span><span class="co">## [2023-11-13 08:08:00.601973] - Removing the ones with fewer than 1 occurrences...</span></span>
<span><span class="co">## [2023-11-13 08:08:00.602268] - Left with 2700 entities</span></span>
<span><span class="co">## [2023-11-13 08:08:00.602319] - Shuffling them...</span></span>
<span><span class="co">## [2023-11-13 08:08:00.603132] Entity type G:</span></span>
<span><span class="co">## [2023-11-13 08:08:00.603183] - Found 16634 entities</span></span>
<span><span class="co">## [2023-11-13 08:08:00.603223] - Removing the ones with fewer than 1 occurrences...</span></span>
<span><span class="co">## [2023-11-13 08:08:00.604510] - Left with 16634 entities</span></span>
<span><span class="co">## [2023-11-13 08:08:00.604562] - Shuffling them...</span></span>
<span><span class="co">## [2023-11-13 08:08:00.609390] Preparing counts and dictionaries for entities and relation types:</span></span>
<span><span class="co">## [2023-11-13 08:08:00.609620] - Writing count of entity type C and partition 0</span></span>
<span><span class="co">## [2023-11-13 08:08:00.610298] - Writing count of entity type G and partition 0</span></span>
<span><span class="co">## [2023-11-13 08:08:00.613050] Preparing edge path /home/vincent/TEMP/RtmpXoctuO/file326126f5c4253/simba_work/pbg/graph0/input/edge, out of the edges found in /home/vincent/TEMP/RtmpXoctuO/file326126f5c4253/simba_work/pbg/graph0/pbg_graph.txt</span></span>
<span><span class="co">## using fast version</span></span>
<span><span class="co">## [2023-11-13 08:08:00.613255] Taking the fast train!</span></span>
<span><span class="co">## [2023-11-13 08:08:00.906598] - Processed 100000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:01.199517] - Processed 200000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:01.493510] - Processed 300000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:01.788421] - Processed 400000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:02.081128] - Processed 500000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:02.372055] - Processed 600000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:02.665790] - Processed 700000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:02.958235] - Processed 800000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:03.251115] - Processed 900000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:03.544382] - Processed 1000000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:03.838060] - Processed 1100000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:04.129736] - Processed 1200000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:04.420804] - Processed 1300000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:04.713416] - Processed 1400000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:05.005179] - Processed 1500000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:05.296857] - Processed 1600000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:05.586542] - Processed 1700000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:05.877504] - Processed 1800000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:06.167762] - Processed 1900000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:06.457530] - Processed 2000000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:06.745587] - Processed 2100000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:07.032731] - Processed 2200000 edges so far...</span></span>
<span><span class="co">## [2023-11-13 08:08:09.205055] - Processed 2286884 edges in total</span></span>
<span><span class="co">## Starting training ...</span></span>
<span><span class="co">## Finished</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gout</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "gdf"           "work_contents" "workdir"       "C_emb"        </span></span>
<span><span class="co">## [5] "G_emb"         "ppCG"</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gout</span><span class="op">$</span><span class="va">C_emb</span> <span class="co"># 50-dimensional embedding of cells</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 50 2700 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): X</span></span>
<span><span class="co">## rownames(50): 0 1 ... 48 49</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(2700): AGCCTCACGTTCGA-1 CTACCTCTCAACCA-1 ... ATCTGTTGCCTTCG-1</span></span>
<span><span class="co">##   CCAATGGAACAGCT-1</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>The components of the output are described in the man page. For this example there is an <em>a priori</em> labeling of each cell recorded in the <code>celltype</code> colData element (that was present in the original h5ad serialization). It needs to be propagated to the embedding.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">gout</span><span class="op">$</span><span class="va">C_emb</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## character(0)</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gout</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate_C_emb.html">annotate_C_emb</a></span><span class="op">(</span><span class="va">gout</span>, <span class="st">"celltype"</span><span class="op">)</span>  <span class="co"># endomorphism</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">gout</span><span class="op">$</span><span class="va">C_emb</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "celltype"</span></span></code></pre>
<p>To support exploration of an embedding without rebuilding it, a serialized embedding is available.</p>
<pre><code><span><span class="va">get_fitted_3k</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="op">)</span> <span class="op">{</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"gout/gout3k.rds"</span>, package<span class="op">=</span><span class="st">"BiocSIMBA"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
<p>The most familiar visualization (of the cell embedding) is based on UMAP.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#, fig.width=unit(8, "cm")}</span></span>
<span><span class="co"># g3k = get_fitted_3k()  # could use</span></span>
<span><span class="fu"><a href="../reference/viz_cemb_umap.html">viz_cemb_umap</a></span><span class="op">(</span><span class="va">gout</span>, <span class="st">"celltype"</span>, ltxtsize<span class="op">=</span><span class="fl">6</span>, ptsize<span class="op">=</span><span class="fl">1</span>, lkeysize<span class="op">=</span><span class="fl">.4</span><span class="op">)</span></span></code></pre></div>
<p><img src="BiocSIMBA_files/figure-html/viz1-1.png" width="700"></p>
<p>The joint embedding can be visualized with:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">jj</span> <span class="op">=</span> <span class="fu"><a href="../reference/joint_emb_CG.html">joint_emb_CG</a></span><span class="op">(</span> <span class="va">gout</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Using the <span style="color: #00BB00;">'X'</span> assay as the <span style="color: #00BB00;">X matrix</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Using the <span style="color: #00BB00;">'X'</span> assay as the <span style="color: #00BB00;">X matrix</span></span></span></code></pre>
<pre><code><span><span class="co">## Performing softmax transformation for query data 0;</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/viz_joint_umap.html">viz_joint_umap</a></span><span class="op">(</span> <span class="va">jj</span>, <span class="st">"celltype"</span>, ltxtsize<span class="op">=</span><span class="fl">6</span>, ptsize<span class="op">=</span><span class="fl">1</span>, lkeysize<span class="op">=</span><span class="fl">.4</span> <span class="op">)</span></span></code></pre></div>
<p><img src="BiocSIMBA_files/figure-html/viz2-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>NULL</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Provided without <strong>any warranty</strong>.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
